concordium-consensus/src/Concordium/Skov/Update.hs

```diff
@@ -478,43 +498,97 @@ doStoreBlock pb@GB.PendingBlock{..} = unlessShutDown $ do
+        processPending slotTime maybeParentBlockSlot = do
+                -- Check:
+                -- - Claimed baker key is valid in committee
+                -- - Proof is valid
+                -- - Signature is correct with the claimed key.
+                -- - Transactions check out OK.
+                lastFin <- fst <$> getLastFinalized
+                lastFinBS <- blockState lastFin
+                gd <- getGenesisData
+                let continuePending = checkClaimedSignature $ do
+                        -- We execute the block transactions in the context of the last finalized block
+                        processBlockTransactions slotTime lastFinBS >>= \case
+                            Nothing -> processDead
+                            Just (newBlock, _) -> addBlockAsPending newBlock
+                getDefiniteSlotBakers gd lastFinBS (blockSlot pb) >>= \case
+                    Just bakers -> case lotteryBaker bakers (blockBaker pb) of
+                        Just (bkrInfo, bkrPower)
+                            | bkrInfo ^. bakerSignatureVerifyKey /= blockBakerKey pb -> do
+                                logEvent Skov LLWarning $ "Pending block is not signed by a valid baker: " ++ show pb
+                                processDead
+                            | otherwise -> do
+                                lastFinSS <- getSeedState lastFinBS
+                                case predictLeadershipElectionNonce lastFinSS (blockSlot lastFin) maybeParentBlockSlot (blockSlot pb) of
+                                    Nothing -> continuePending -- Cannot check the proof (yet)
+                                    Just nonceList -> do
+                                        -- We get the election difficulty based on the last finalized block.
+                                        -- In principle, this could change before the block.
+                                        elDiff <- getElectionDifficulty lastFinBS slotTime
+                                        let verifyProofWithNonce nonce = verifyProof
+                                                nonce
+                                                elDiff
+                                                (blockSlot pb)
+                                                (bkrInfo ^. bakerElectionVerifyKey)
+                                                bkrPower
+                                                (blockProof pb)
+                                        if any verifyProofWithNonce nonceList
+                                        then do
+                                            continuePending
+                                        else do
+                                            logEvent Skov LLWarning $ "Block proof of pending block was not verifiable: " ++ show pb
+                                            processDead
+                        Nothing -> processDead
+                    Nothing -> continuePending
```

Comment:
Is it an explicit choice here to do the signature check at the very end of processing?